File: Top_Google_Questions/1110._Delete_Nodes_And_Return_Forest/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def delNodes(self, root: TreeNode, to_delete: List[int]) -> List[TreeNode]:
        ans = []
        s = set(to_delete)
        
        def helper(root, parent, isLeftChild):
            if not root:
                return
            
            helper(root.left, root, True)
            helper(root.right, root, False)
            
            if root.val in s:
                if isLeftChild:
                    parent.left = None
                else:
                    parent.right = None
            
                if root.left:
                    ans.append(root.left)

                if root.right:
                    ans.append(root.right)
        
        if root.val not in s:
            ans.append(root)
        
        dummyRoot = TreeNode()
        dummyRoot.left = root
        
        helper(root, dummyRoot, True)
        del dummyRoot
            
        return ans
            
        

File: Top_Google_Questions/1110._Delete_Nodes_And_Return_Forest/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number[]} to_delete
 * @return {TreeNode[]}
 */
var delNodes = function(root, to_delete) {
    let ans = [];
    let set = {};
    to_delete.forEach(v => set[v] = 0);
    
    let helper = function(root, parent, isLeftChild) {
        if (!root) return;
        
        helper(root.left, root, true);
        helper(root.right, root, false);
        
        if (set.hasOwnProperty(root.val)) {
            if (isLeftChild)
                parent.left = null;
            else
                parent.right = null;
            
            if (root.left)
                ans.push(root.left);
            
            if (root.right)
                ans.push(root.right);
        }
    }
    
    if (!set.hasOwnProperty(root.val))
        ans.push(root);
    
    let dummyParent = new TreeNode(0, root);
    helper(root, dummyParent, true);
    
    return ans;
};